### PR TITLE
Fix BF16 teardown for TPU precision plugin

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

The env variable gets set in the tpu bf16 precision plugin `connect()` method but gets deleted again in the TPUSpawnPlugin before training starts. The correct behavior is to set the variable at the beginning of a run and remove it again at the end of a run (teardown). For this reason, we are introducing a `teardown()` method on the `PrecisionPlugin` base class (missing so far due to lack of use case). 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs) #10896 
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda @kaushikb11

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 1]
```diff
@@ -243,10 +243,6 @@ def get_mp_spawn_kwargs(self, trainer: Optional["pl.Trainer"] = None) -> Dict[st
         }
 
     def spawn(self, function: Callable, *args: Any, **kwargs: Any) -> Optional[Union[Any, "_SpawnOutput"]]:
-        # TODO: this todo is unclear, does it still apply?
-        # todo: precision pluging is call in accelerator setup and should be moved
-        if "XLA_USE_BF16" in os.environ:
-            del os.environ["XLA_USE_BF16"]
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:249]: @kaushikb11 is this something that should be set up and torn down per run? https://github.com/PyTorchLightning/pytorch-lightning/blob/aeb0b5595fd73d086f4ae0f99d3f1f112f6a4c29/pytorch_lightning/plugins/precision/tpu_bf16.py#L31

otherwise doesn't this break?
```py

t16 = Trainer(accelerator="tpu", precision="bf16")
t16.fit(model)
t = Trainer(accelerator="tpu")
t.fit(model)
```
since the latter could inadvertently re-use the environment variable?

Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:249]: Agreed, it is a bug. This definitely needs to be deleted per run.

Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:249]: this could be done with the `setup_environment` + `teardown` methods of the training type plugin

**Review by rohitgr7 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: tests/models/test_tpu.py [hunk 2]
```diff
@@ -144,7 +143,6 @@ def test_model_16bit_tpu_index(tmpdir, tpu_core):
     model = BoringModel()
     tpipes.run_model_test(trainer_options, model, on_gpu=False)
     assert torch_xla._XLAC._xla_get_default_device() == f"xla:{tpu_core}"
-    assert os.environ.get("XLA_USE_BF16") == str(1), "XLA_USE_BF16 was not set in environment variables"
```
Comment (carmocca) [inline tests/models/test_tpu.py:147]: Should this PR also remove: https://github.com/PyTorchLightning/pytorch-lightning/blob/eec2bae6c984895d1d4df4e75cf628c58128928a/pytorch_lightning/plugins/training_type/tpu_spawn.py#L246-L248
?

File: CHANGELOG.md [hunk 3]
```diff
@@ -55,6 +55,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added `console_kwargs` for `RichProgressBar` to initialize inner Console ([#10875](https://github.com/PyTorchLightning/pytorch-lightning/pull/10875))
 
 
+- Added a `PrecisionPlugin.teardown` method ([#????](https://github.com/PyTorchLightning/pytorch-lightning/issues/????))
```
Comment (tchaton) [inline CHANGELOG.md:58]:
```suggestion
- Added a `PrecisionPlugin.teardown` method ([#10990](https://github.com/PyTorchLightning/pytorch-lightning/issues/10990))
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -331,6 +335,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fixed double evaluation bug with fault-tolerance enabled where the second call was completely skipped ([#11119](https://github.com/PyTorchLightning/pytorch-lightning/pull/11119))
 
+
+- Fixed an issue with the `TPUSpawnPlugin` handling the `XLA_USE_BF16` environment variable incorrectly ([#????](https://github.com/PyTorchLightning/pytorch-lightning/pull/????))
```
Comment (tchaton) [inline CHANGELOG.md:339]:
```suggestion
- Fixed an issue with the `TPUSpawnPlugin` handling the `XLA_USE_BF16` environment variable incorrectly ([#10990](https://github.com/PyTorchLightning/pytorch-lightning/pull/10990))
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 5]
```diff
@@ -425,6 +425,7 @@ def teardown(self) -> None:
 
         It is the right place to release memory and free other resources.
         """
+        self.precision_plugin.teardown()
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:447]: We should remove the `abstractmethod` there IMO.

File: tests/models/test_tpu.py (hunk 2 above)
Comment (tchaton) [reply tests/models/test_tpu.py:147]: I believe so.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:446]: I made sure to call super() and remove the abstractmethod decorator (setup also doesn't have it)

**Review by Borda — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]
